[Logging] Report whether video and audio tracks are protected in logging and in Web Inspector

jernoble opened the pull request:
#### 3ff93e0dd934c5319ed32743636e5320e674993f
<pre>
[Logging] Report whether video and audio tracks are protected in logging and in Web Inspector
<a href="https://rdar.apple.com/165026619">rdar://165026619</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302768">https://bugs.webkit.org/show_bug.cgi?id=302768</a>

Reviewed by Devin Rousso.

Use the CMFormatDescription ProtectedContentOriginalFormat extention to determine whether
the given track has protected contents, and report that data in logging, including via
Web Inspector.

Drive-by fix: Audio tracks were not having their configuration updated by AVTrackPrivateAVFObjCImpl.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::descriptionForTrack):
* Source/WebCore/html/track/AudioTrackConfiguration.cpp:
(WebCore::AudioTrackConfiguration::toJSON const):
* Source/WebCore/html/track/AudioTrackConfiguration.h:
(WebCore::AudioTrackConfiguration::isProtected const):
(WebCore::AudioTrackConfiguration::setProtected):
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::setProtected):
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::isProtected const):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/platform/graphics/PlatformAudioTrackConfiguration.h:
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::audioTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::isProtected const):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WebCore::codecFromFormatDescription):
(WebCore::formatDescriptionIsProtected):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/avfoundation/objc/AudioTrackPrivateMediaSourceAVFObjC.cpp:
(WebCore::AudioTrackPrivateMediaSourceAVFObjC::resetPropertiesFromTrack):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
(WI.MediaDetailsSidebarPanel):
(WI.MediaDetailsSidebarPanel.prototype.layout):
(WI.MediaDetailsSidebarPanel.prototype.setVideo):
(WI.MediaDetailsSidebarPanel.prototype.setAudio):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/303255@main">https://commits.webkit.org/303255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b72052e9f42458599aabb442166be2112c45a74

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139324 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133683 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4067 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100748 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134759 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118055 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81536 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2929 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82544 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123876 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141968 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130320 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3974 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109123 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4055 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3481 "Found 2 new test failures: http/tests/webgpu/webgpu/api/operation/sampling/filter_mode.html imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/white-space_normal_wrapped.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109287 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27686 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3018 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114336 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57184 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4028 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32738 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163286 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3860 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67475 "Built successfully") | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4120 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3988 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -317,14 +321,16 @@ WI.MediaDetailsSidebarPanel = class MediaDetailsSidebarPanel extends WI.DOMDetai
         this.#ui.resolutionRow.pendingValue = WI.UIString("%dx%d (%dfps)").format(video?.width ?? 0, video?.height ?? 0, video?.framerate ?? 0);
         this.#ui.resolutionRow.element.classList.toggle("hidden", !video);
         this.#ui.videoCodecRow.pendingValue = video?.codec;
+        this.#ui.videoProtectedRow.pendingValue = video?.isProtected ? WI.UIString("True") : null;
```
the word "True" is a bit odd.  id suggest using ✔ (a.k.a. `"\u2713"`) like Web Inspector does in other places similar to this (e.g. cookie "Secure")

dcrousso commented on Source/WebCore/html/MediaElementSession.cpp:
```diff
@@ -1671,6 +1671,8 @@ void MediaElementSession::clientCharacteristicsChanged(bool positionChanged)
 String MediaElementSession::descriptionForTrack(const AudioTrack& track)
 {
+    if (track.configuration().isProtected())
+        return makeString(track.configuration().codec(), " protected"_s);
```
rather than adding a separate "Protected" row should we just format the existing "Video Codec" row in a similar way (e.g. `"%s (protected)"`)?

jernoble replied:
There's too many edge cases here; `codec` itself is optional, so it becomes a branching mess of `typeof codec === "string" && codec.length > 0` checks everywhere. This code above produces an intentionally compact description. There's much more room to be ... expressive in the Web Inspector.

dcrousso commented on Source/JavaScriptCore/inspector/protocol/DOM.json:
```diff
@@ -223,7 +223,8 @@
                 { "name": "numberOfChannels", "type": "integer", "description": "The number of audio channels in the primary audio track." },
-                { "name": "sampleRate", "type": "number", "description": "The sample rate of the primary audio track in hertz." }
+                { "name": "sampleRate", "type": "number", "description": "The sample rate of the primary audio track in hertz." },
+                { "name": "isProtected", "type": "boolean", "description": "Whether the track contains protected contents" }
```
id recommend making this optional since it seems like it not being provided is indistinguishable from `false` (i.e. avoid sending unnecessary data)

dcrousso commented on Source/JavaScriptCore/inspector/protocol/DOM.json:
```diff
@@ -239,7 +240,8 @@
                 { "name": "spatialVideoMetadata", "$ref": "SpatialVideoMetadata", "optional": true },
-                { "name": "videoProjectionMetadata", "$ref": "VideoProjectionMetadata", "optional": true }
+                { "name": "videoProjectionMetadata", "$ref": "VideoProjectionMetadata", "optional": true },
+                { "name": "isProtected", "type": "boolean", "description": "Whether the track contains protected contents" }
```
ditto

if for whatever reason you dont want to make this optional then please move this above the other optional properties as we try to keep optional properties last for better readability at a glance

dcrousso commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -3267,6 +3267,7 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::DOM::MediaStats>> In
             .setFramerate(configuration.framerate())
             .setHeight(configuration.height())
             .setWidth(configuration.width())
+            .setIsProtected(configuration.isProtected())
```
Web Inspector normally doesnt add optional JSON when the value is the same as the default
```cpp
if (configuration.isProtected())
    videoJSON->setIsProtected("isProtected", true);
```

dcrousso commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -3295,6 +3296,7 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::DOM::MediaStats>> In
             .setHumanReadableCodecString(humanReadableStringFromCodecString(configuration.codec()))
             .setNumberOfChannels(configuration.numberOfChannels())
             .setSampleRate(configuration.sampleRate())
+            .setIsProtected(configuration.isProtected())
```
ditto

dcrousso reviewed (approved):
neat!

is there a way to add a test for this?

jernoble committed:
[Logging] Report whether video and audio tracks are protected in logging and in Web Inspector
rdar://165026619
https://bugs.webkit.org/show_bug.cgi?id=302768

Reviewed by Devin Rousso.

Use the CMFormatDescription ProtectedContentOriginalFormat extention to determine whether
the given track has protected contents, and report that data in logging, including via
Web Inspector.

Drive-by fix: Audio tracks were not having their configuration updated by AVTrackPrivateAVFObjCImpl.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::descriptionForTrack):
* Source/WebCore/html/track/AudioTrackConfiguration.cpp:
(WebCore::AudioTrackConfiguration::toJSON const):
* Source/WebCore/html/track/AudioTrackConfiguration.h:
(WebCore::AudioTrackConfiguration::isProtected const):
(WebCore::AudioTrackConfiguration::setProtected):
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::setProtected):
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::isProtected const):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/platform/graphics/PlatformAudioTrackConfiguration.h:
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::audioTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::isProtected const):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WebCore::codecFromFormatDescription):
(WebCore::formatDescriptionIsProtected):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/avfoundation/objc/AudioTrackPrivateMediaSourceAVFObjC.cpp:
(WebCore::AudioTrackPrivateMediaSourceAVFObjC::resetPropertiesFromTrack):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
(WI.MediaDetailsSidebarPanel):
(WI.MediaDetailsSidebarPanel.prototype.layout):
(WI.MediaDetailsSidebarPanel.prototype.setVideo):
(WI.MediaDetailsSidebarPanel.prototype.setAudio):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/303255@main
